### PR TITLE
Ensure reset of deferred validation

### PR DESCRIPTION
**Pull request by dpep**
I was digging through this code and noticed a few things that could be tightened up (it's generally an awesome library though, so thanks!).  Forgive me if this is unwelcomed feedback.

There is an edge case where the block in defer_validation could raises an exception and validation_deferred isn't reset properly.  Perhaps contrived, but I thought it worth addressing.  The rest are more cosmetic suggestions to make this code simpler / more Rubyish

### CODE REVIEW

File: lib/addressable/uri.rb [hunk 1]
```diff
@@ -851,7 +849,8 @@ def initialize(options={})
         self.query_values = options[:query_values] if options[:query_values]
         self.fragment = options[:fragment] if options[:fragment]
       end
-      self.to_s
+
+      to_s # enforce path validation
```
Comment (dpep) [inline lib/addressable/uri.rb:853]: this line took me a bit to figure out, so I figured it was comment worthy.    based on the specs, it looks like `initialize` leverages the validation of `self.path` done in `.to_s`.  but moving it into `.validate` causes other test failures...

**Comment by dentarg** [comment]
>The rest are more cosmetic suggestions to make this code simpler / more Rubyish

Can you put those in a separate PR?

**Comment by dpep** [comment]
@dentarg : spun out into https://github.com/sporkmonger/addressable/pull/482 per your request

File: spec/addressable/uri_spec.rb [hunk 2]
```diff
@@ -6743,3 +6743,25 @@ def to_str
...
+
+  let(:uri) { Addressable::URI.parse("http://example.com") }
+
+  it "defers validation within the block" do
+    uri.defer_validation do
+      expect(deferred).to be true
+    end
+  end
+
+  it "always resets deferal afterward" do
```
Comment (dentarg) [inline spec/addressable/uri_spec.rb:6758]:
```suggestion
  it "always resets deferral afterward" do
```

**Pull request merged**